src/java/org/apache/cassandra/FilterExperiment.java

```diff
@@ -147,10 +150,67 @@ private static Iterator<Cell> iterator(ColumnFamily columnFamily) {
                                 Predicates.not(columnFamily.inOrderDeletionTester()::isDeleted));
     }
 
+    private static Optional<Cell> getNext(Iterator<Cell> iterator)
+    {
+        return iterator.hasNext() ? Optional.ofNullable(iterator.next()) : Optional.empty();
+    }
+
     static boolean areTrulyEqual(ColumnFamily legacy, ColumnFamily modern) {
         return ColumnFamily.digest(legacy).equals(ColumnFamily.digest(modern));
     }
 
+    static ComparisonResult compareByIterator(Iterator<Cell> legacyIterator, Iterator<Cell> modernIterator)
+    {
+        ImmutableList.Builder<String> differencesBuilder = ImmutableList.builder();
+        int index = -1;
+        while (legacyIterator.hasNext() || modernIterator.hasNext())
+        {
+            index++;
+            Optional<Cell> legacy = getNext(legacyIterator);
+            Optional<Cell> modern = getNext(modernIterator);
+
+            if (legacy.isPresent() && modern.isPresent())
+            {
+                if (legacy.get().equals(modern.get()))
+                {
+                    continue;
+                }
+                differencesBuilder.add(notEqual(index, legacy.get(), modern.get()));
+            }
+            else
+            {
+                differencesBuilder.add(missingItem(index, legacy.isPresent(), modern.isPresent()));
+            }
+        }
+        List<String> differences = differencesBuilder.build();
+
+        if (differences.isEmpty()) {
+            return ComparisonResult.EQUAL;
+        } else {
+            log.warn("Column families returned different results via iterator: {}", SafeArg.of("differences", differences));
+            return ComparisonResult.NOT_EQUAL_BY_ITERATOR;
+        }
+    }
+
+    private static String missingItem(int index, boolean hasLegacyCell, boolean hasModernCell) {
+        return String.format("[%s: MissingItem(hasLegacyCell=%s, hasModernCell=%s)]",
```

Comment:
String.format will be expensive to parse and allocate on potentially hot comparison paths

Reply:
This should not be a hot comparison path, we see maybe 5-10 failures per month across the fleet. Since we only run the String.format on failures we should be good? Given this is so hard to reproduce I think we should prioritize getting as much information as possible.

Definitely open to cheaper ways of logging though.